src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs

```diff
@@ -24,33 +65,82 @@ private static long ParseContentLength(string value)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private void SetValueUnknown(string key, in StringValues value)
+        private void AppendContentLength(Span<byte> value)
         {
-            Unknown[key] = value;
+            if (_contentLength.HasValue)
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.MultipleContentLengths);
+            }
+
+
+            if (!TryParseNonNegativeInt64(value, out var parsed))
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidContentLength, value.GetAsciiOrUTF8StringNonNullCharacters());
+            }
+
+            _contentLength = parsed;
         }
 
-        public unsafe void Append(Span<byte> name, string value)
+        private static unsafe bool TryParseNonNegativeInt64(Span<byte> value, out long result)
         {
-            fixed (byte* namePtr = name)
```

Comment:
Should we use the Utf8Parser here?

Reply:
Not sure https://github.com/dotnet/corefx/issues/35814

Reply:
Also mostly a copy from [aspnet/AspNetCore/src/Http/Headers/src/HeaderUtilities.cs](https://github.com/aspnet/AspNetCore/blob/285110de91f276d0b9302c3d2a84f0dc5af861b1/src/Http/Headers/src/HeaderUtilities.cs#L422-L456) but changed to operate over a `Span<byte>` rather than a `StringSegment`.

Reply:
We should add an overload in the same place:

```c#
public static unsafe bool TryParseNonNegativeInt64(StringSegment value, out long result);
public static unsafe bool TryParseNonNegativeInt64(ReadOnlySpan<byte> span, out long result);
```

Reply:
Moved